codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3UpdateEndpoint.java

```diff
@@ -0,0 +1,154 @@
+                            AwsCustomGoDependency.S3_CUSTOMIZATION).build(),
+                    SymbolUtils.createValueSymbolBuilder(UPDATE_ENDPOINT_INTERNAL_ADDER + "Options",
+                            AwsCustomGoDependency.S3_CUSTOMIZATION).build(),
+                    USE_PATH_STYLE_OPTION,
+                    USE_ACCELERATE_OPTION
+            );
+        });
+        writer.insertTrailingNewline();
+    }
+
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return ListUtils.of(
+                // Add S3 config to use path style host addressing.
+                RuntimeClientPlugin.builder()
+                        .servicePredicate(S3UpdateEndpoint::isS3Service)
+                        .configFields(ListUtils.of(
+                                ConfigField.builder()
+                                        .name(USE_PATH_STYLE_OPTION)
+                                        .type(SymbolUtils.createValueSymbolBuilder("bool")
+                                                .putProperty(SymbolUtils.GO_UNIVERSE_TYPE, true)
+                                                .build())
+                                        .documentation("Allows you to enable the client to use path-style addressing, "
+                                                + "i.e., `http://s3.amazonaws.com/BUCKET/KEY`. By default, the S3 client"
+                                                + "will use virtual hosted bucket addressing when possible"
+                                                + "(`http://BUCKET.s3.amazonaws.com/KEY`).")
+                                        .build(),
+                                ConfigField.builder()
+                                        .name(USE_ACCELERATE_OPTION)
+                                        .type(SymbolUtils.createValueSymbolBuilder("bool")
+                                                .putProperty(SymbolUtils.GO_UNIVERSE_TYPE, true)
+                                                .build())
+                                        .documentation("Set this to `true` to enable S3 Accelerate feature. For all operations "
+                                                + "compatible with S3 Accelerate will use the accelerate endpoint for "
+                                                + "requests. Requests not compatible will fall back to normal S3 requests. "
+                                                + ""
+                                                + "The bucket must be enable for accelerate to be used with S3 client with "
+                                                + "accelerate enabled. If the bucket is not enabled for accelerate an error "
+                                                + "will be returned. The bucket name must be DNS compatible to also work "
+                                                + "with accelerate."
+                                        )
+                                        .build()
```

Comment:
Probably want to leave accelerate out of this for now since it adds complexity. Will help focus on just virtual host addressing.